dftimewolf/lib/collectors/gcp_logging.py

```diff
@@ -42,75 +40,121 @@ def __init__(self,
+    results = logging_client.list_entries(  # type: ignore
+          order_by=logging.DESCENDING,
+          filter_=self._filter_expression,
+          page_size=1000)
+    return results.pages
+
+  def ProcessPages(self, pages, backoff_multiplier, output_file, output_path):
+    """Iterates through a generator or pages and saves logs to disk.
+    Can optionally perform exponential backoff if query API limits are exceeded.
+    
+    Args:
+      pages (generator): A google cloud logging list_entries generator pages object
+      backoff_multiplier (str): Query delay multiplier if the API quota is met and backoff is enabled
+      output_file (str): Output file name
+      output_path (str): Output file path
+    
+    Returns:
+      output_path (str): Log output path (may have been updated if API quota was exceeded)
+    """
+    while True:
+      try:
+        time.sleep(self._delay)
+        page = next(pages)
+      except google_api_exceptions.TooManyRequests as exception:
+        self.PublishMessage('Hit quota limit requesting GCP logs.')
+        if self._backoff is True:
+          self.PublishMessage('Retrying in 60 seconds with a slower query rate.')
+          self.PublishMessage('Due to the GCP logging API, the query must restart from the beginning')
+          time.sleep(60)
+          if self._delay == 0:
+            self._delay = 1
+          else:
+            self._delay *= backoff_multiplier
+          self.PublishMessage('Setting up new logging client.')
+          logging_client = self.SetupLoggingClient()
+          pages = self.ListPages(logging_client)
+          self.PublishMessage('Restarting query with an API request rate of 1 per {0}s'.format(self._delay))
+          output_file, output_path = self.OutputFile()
+        else:
+          self.PublishMessage('Exponential backoff was not enabled, so query has exited.')
+          self.PublishMessage('The collection is most likely incomplete.')
+      except StopIteration:
+        break
+
+      for entry in page:
+        log_dictionary = entry.to_api_repr()
+        output_file.write(json.dumps(log_dictionary))
+        output_file.write('\n')
+    
+    return output_path
 
   # pylint: disable=arguments-differ
-  def SetUp(self, project_name: str, filter_expression: str) -> None:
+  def SetUp(self, project_name: str, filter_expression: str, backoff: bool, delay: str) -> None:
     """Sets up a a GCP logs collector.
 
     Args:
       project_name (str): name of the project to fetch logs from.
       filter_expression (str): GCP advanced logs filter expression.
+      backoff (bool): Retry queries with an increased delay when API quotas are exceeded.
+      delay (str): Seconds to wait between retreiving results pages to avoid exceeding API quotas
     """
     self._project_name = project_name
     self._filter_expression = filter_expression
+    self._backoff = backoff
+    self._delay = float(delay)
+    self.PublishMessage(f'backoff {self._backoff}')
```

Comment:
Similar to above on `PublishMessage` - I don't think this output is necessary, user should know what values they just put in, and for debugging and module params are included in debug output already.

Reply:
I ended up removing this message completely. After initially changing it to debug logging, I realised this is redundant as dftimewolf logs the arguments (backoff and delay) configuration to debug on execution.